File: namhee/perm-comb/boj15650.java
import java.util.Scanner;

public class Main {
    static int N, M;
    static int[] arr;
    static boolean[] visited;
    
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        
        arr = new int[M];
        visited = new boolean[N + 1];
        perm(1, 0);
    }
    
    static void perm(int start, int depth) {
        if (depth == M) {
            StringBuilder sb = new StringBuilder("");
            for (int i = 0; i < M; i++) {
                sb.append(arr[i] + " ");
            }
            System.out.println(sb);
            return;
        }
        
        for (int i = start; i <= N; i++) {
            if (!visited[i]) {
                arr[depth] = i;
                visited[i] = true;
                perm(i + 1, depth + 1);
                visited[i] = false;
            }
        }
    }
}

File: sangwook/simulation/boj15650.kt
package src.main.kotlin.simulation

import java.lang.StringBuilder

class boj15650 {

    fun sol() {
        val (n, m) = readLine()!!.split(" ").map { it.toInt() }
        val arr = IntArray(m)
        comb(0, 0, n, m, arr)
    }

    fun comb(cur: Int, start: Int, n: Int, m: Int, arr: IntArray) {

        if (cur == m) {
            val sb = StringBuilder()
            arr.forEach { sb.append(it).append(' ') }
            println(sb.toString())
            return
        }

        for (i in start until n) {
            arr[cur] = i + 1
            comb(cur + 1, i + 1, n, m, arr)
        }

    }
}

fun main() {
    boj15650().sol()
}